omi/sawtooth_omi/handler.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2017 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+import hashlib
+
+from sawtooth_sdk.processor.state import StateEntry
+from sawtooth_sdk.processor.exceptions import InvalidTransaction
+from sawtooth_sdk.processor.exceptions import InternalError
+from sawtooth_sdk.protobuf.transaction_pb2 import TransactionHeader
+
+from sawtooth_omi.protobuf.work_pb2 import Work
+from sawtooth_omi.protobuf.recording_pb2 import Recording
+from sawtooth_omi.protobuf.identity_pb2 import IndividualIdentity, OrganizationalIdentity
+from sawtooth_omi.protobuf.txn_payload_pb2 import OMITransactionPayload
+
+
+# actions
+
+WORK = 'SetWork'
+RECORDING = 'SetRecording'
+INDIVIDUAL = 'SetIndividualIdentity'
+ORGANIZATION = 'SetOrganizationalIdentity'
+
+# address
+
+def _hash_name(name):
+    return hashlib.sha512(name.encode('utf-8')).hexdigest()
+
+
+FAMILY_NAME = 'OMI'
+OMI_ADDRESS_PREFIX = _hash_name(FAMILY_NAME)[:6]
+
+
+def _get_address_infix(action):
+    infixes = {
+        WORK: 'a0',
+        RECORDING: 'a1',
+        INDIVIDUAL: '00',
+        ORGANIZATION: '01',
+    }
+
+    return infixes[action]
+
+
+def _get_unique_key(obj, action):
+    if action in (WORK, RECORDING):
+        key = obj.title
+    elif action in (INDIVIDUAL, ORGANIZATION):
+        key = obj.name
+
+    return key
+
+
+def make_omi_address(obj, action):
+    infix = _get_address_infix(action)
+    key = _get_unique_key(obj, action)
+
+    return OMI_ADDRESS_PREFIX + infix + _hash_name(key)[-62]:
+
+
+class OMITransactionHandler:
+    @property
+    def family_name(self):
+        return FAMILY_NAME
+
+    @property
+    def family_versions(self):
+        return ['1.0']
+
+    @property
+    def encodings(self):
+        return ['application/protobuf']
+
+    @property
+    def namespaces(self):
+        return [OMI_ADDRESS_PREFIX]
+
+    def apply(self, transaction, state):
+        action, txn_obj, signer = _unpack_transaction(transaction)
+
+        address = make_omi_address(txn_obj, action)
+
+        state_obj = _get_state_object(state, address, action)
+
+        if state_obj:
+            _validate_object(state_obj, action, signer)
```

Comment:
This is less about validating the object, which there should probably be some (minimal, at the moment) validation of the transaction payload, and more about authorizing updates.

I.e. if not `state_obj`, it's an insert, else update.